### ISSUE TITLE
Some kinds of ERB commenting break highlighting

### ISSUE THREAD

**Issue by csuhta** [atom/language-ruby#263]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Certain kinds of ERB commenting will break the syntax highlighting flow, such as `<% # comment %>` and `<%-# comment %>`

### Steps to Reproduce

A poorly colorized file shown below:

```erb
<%# This is a ERB comment %>
<%= @object.variable %>

<%-# This is also a ERB comment %> <!-- Styles broken past this point --->
<%= @object.variable %>

<%# This is a comment ERB %>
<%= @object.variable %>

<% # This is a comment %> <!-- Styles broken past this point --->
<%= @object.variable %> 
```

<img width="701" alt="q0rwisizvd" src="https://user-images.githubusercontent.com/769083/53940881-bb13ab80-4084-11e9-92fd-ee7094261805.png">

<img width="673" alt="3kcntrvorq" src="https://user-images.githubusercontent.com/769083/53940891-bea73280-4084-11e9-8fca-c6e5eca014e4.png">

**Expected behavior:** Syntax highlighting handles the comment

**Actual behavior:** Syntax highlighting breaks down for future Ruby blocks after the comment

*Edit by @rsese to add screenshot*

![erb-comment-highlighting](https://github-slack.s3.amazonaws.com/monosnap/test2.html.erb__worktest-tree-sitter_2019-03-12_01-58-13.png)

**Reproduces how often:** Always

### Versions

*Edit by @rsese: also reproduced with 1.37.0-nightly12 on macOS 10.12.*

```
❯❯ apm --version
apm  2.1.2
npm  6.2.0
node 8.9.3 x64
atom 1.33.1
python 2.7.15
git 2.20.1

❯❯ atom --version
Atom    : 1.33.1
Electron: 2.0.16
Chrome  : 61.0.3163.100
Node    : 8.9.3
```

**Comment by rsese** [comment]
Thanks for the report!  Reproduced with 1.37.0-nightly12 on macOS 10.12.6.

![erb-comment-highlighting](https://github-slack.s3.amazonaws.com/monosnap/test2.html.erb__worktest-tree-sitter_2019-03-12_01-58-13.png)

**Comment by nebulousdog** [comment]
confirmed that this is a `tree-sitter` issue, so I turned this back off and no longer a zillion errors:

![image](https://user-images.githubusercontent.com/2218331/56779289-262d5480-678f-11e9-9d60-a580e6309538.png)

```
> apm --version
apm  2.1.3
npm  6.2.0
node 8.9.3 x64
atom 1.36.0-beta1
python 2.7.15+
git 2.19.1

> lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.10
Release:	18.10
Codename:	cosmic
```

**Comment by jasonrudolph** [comment]
> **Actual behavior:** Syntax highlighting breaks down for future Ruby blocks after the comment

☝️ This is a nice summary of the issue.  Thanks, @csuhta.

@maxbrunsfeld and I looked into this issue yesterday. Building on the summary above, the syntax highlighting works correctly when using an ERB-style comment (i.e., `<%#`), but it breaks when using an ERB statement that *includes* a Ruby-style comment (i.e., a `<% `, `<%- `, or `<%= ` node that _contains_ a Ruby comment ):

<img width="862" alt="Screen Shot 2019-05-03 at 9 48 36 AM" src="https://user-images.githubusercontent.com/2988/57142345-4a3df680-6d8a-11e9-8238-a713e697515c.png">

We found that this issue also exists for EJS:

<img width="1057" alt="Screen Shot 2019-05-03 at 9 55 15 AM" src="https://user-images.githubusercontent.com/2988/57142346-4a3df680-6d8a-11e9-86fd-a83405e27a56.png">

It seems like fixing this issue is going to be a rather substantial undertaking, so I don't yet know when we'll be able to address this issue. In the meantime, if you're able to use ERB-style comments exclusively (which I know won't be a viable solution for everyone), then you can work around this bug.

We'll likely open an issue with additional details from our findings, since the underlying issue goes beyond language-ruby and ERB.

**Comment by jasonrudolph** [comment]
**tl;dr**: https://github.com/atom/atom/pull/19291 resolves the most common cases where this issue crops us, and the fix is targeted for inclusion in Atom 1.38. 😅

## Before and After

### Ruby comments

#### Before

<img width="911" alt="Screen Shot 2019-05-09 at 10 05 03 AM" src="https://user-images.githubusercontent.com/2988/57459815-f3c93000-7241-11e9-9117-b1181bf5f738.png">

#### After

<img width="908" alt="Screen Shot 2019-05-09 at 10 05 06 AM" src="https://user-images.githubusercontent.com/2988/57459816-f3c93000-7241-11e9-836b-940ec89b9a64.png">

### ERB comments

#### Before

<img width="911" alt="Screen Shot 2019-05-09 at 10 04 33 AM" src="https://user-images.githubusercontent.com/2988/57459812-f3c93000-7241-11e9-917d-2018cf7f0854.png">

#### After

<img width="908" alt="Screen Shot 2019-05-09 at 10 07 00 AM" src="https://user-images.githubusercontent.com/2988/57459963-3d197f80-7242-11e9-97f6-5086ab5ad36f.png">

## Remaining edge case and available workarounds

In the fairly uncommon case where you have a Ruby comment in an ERB directive followed by additional ERB directives on the same line, the latter ERB directives are currently styled incorrectly:

<img width="908" alt="Screen Shot 2019-05-09 at 10 27 14 AM" src="https://user-images.githubusercontent.com/2988/57461601-507a1a00-7245-11e9-948b-81fd5179aeed.png">

We're tracking this bug in https://github.com/tree-sitter/tree-sitter/issues/327. As seen in the screenshot above, there are two workarounds for this bug in the meantime:

- Option 1: Add a newline after any ERB directives that use Ruby comments, or
- Option 2: Use ERB comments instead of Ruby comments

Because the problem identified in the [issue body above](https://github.com/atom/language-ruby/issues/263#issue-418175953) is now resolved, and because we're tracking the remaining edge case in a [separate issue](https://github.com/tree-sitter/tree-sitter/issues/327), I'm going to close this issue out.

Thanks again for reporting this, @csuhta. :bow:
